File: packages/python/onnx9000-core/src/onnx9000/core/models/vit.py
"""Module docstring."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, concat, flatten, transpose
from onnx9000.core.ops import slice as slice_op
from onnx9000.core.primitives import (
    ConvND,
    Gelu,
    Gemm,
    LayerNormalization,
    MultiHeadAttention,
)


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    return Variable(name=name, shape=shape, dtype=dtype)


class PatchEmbed:
    """Image to Patch Embedding."""

    def __init__(  # noqa: D107
        self,
        img_size: int = 224,
        patch_size: int = 16,
        in_chans: int = 3,
        embed_dim: int = 768,
        prefix: str = "",
    ):
        self.prefix = prefix
        self.img_size = img_size
        self.patch_size = patch_size
        self.num_patches = (img_size // patch_size) * (img_size // patch_size)
        self.proj = ConvND(2, in_chans, embed_dim, kernel_size=patch_size, stride=patch_size)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        x = self.proj(
            x,
            get_param(
                f"{self.prefix}.proj.weight",
                [self.proj.out_channels, self.proj.in_channels, self.patch_size, self.patch_size],
            ),
            get_param(f"{self.prefix}.proj.bias", [self.proj.out_channels]),
        )
        # flatten: [B, C, H, W] -> [B, C, H*W]
        from onnx9000.core.ops import record_op

        x = record_op("Flatten", [x], {"axis": 2})
        # transpose to [B, N, C] (ONNX transpose uses perm)
        x = transpose(x, perm=[0, 2, 1])
        return x


class Block:
    """Transformer Block."""

    def __init__(  # noqa: D107
        self,
        dim: int,
        num_heads: int,
        mlp_ratio: float = 4.0,
        qkv_bias: bool = False,
        prefix: str = "",
    ):
        self.prefix = prefix
        self.dim = dim
        self.norm1 = LayerNormalization((dim,), epsilon=1e-6)
        self.attn = MultiHeadAttention(num_heads=num_heads, qkv_bias=qkv_bias)
        self.norm2 = LayerNormalization((dim,), epsilon=1e-6)
        mlp_hidden_dim = int(dim * mlp_ratio)
        self.fc1 = Gemm(trans_b=1)
        self.act = Gelu()
        self.fc2 = Gemm(trans_b=1)
        self.mlp_hidden_dim = mlp_hidden_dim

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        identity = x
        x = self.norm1(
            x,
            get_param(f"{self.prefix}.norm1.weight", [self.dim]),
            get_param(f"{self.prefix}.norm1.bias", [self.dim]),
        )
        x = self.attn(x, x, x)
        x = add(x, identity)

        identity = x
        x = self.norm2(
            x,
            get_param(f"{self.prefix}.norm2.weight", [self.dim]),
            get_param(f"{self.prefix}.norm2.bias", [self.dim]),
        )
        x = self.fc1(
            x,
            get_param(f"{self.prefix}.fc1.weight", [self.mlp_hidden_dim, self.dim]),
            get_param(f"{self.prefix}.fc1.bias", [self.mlp_hidden_dim]),
        )
        x = self.act(x)
        x = self.fc2(
            x,
            get_param(f"{self.prefix}.fc2.weight", [self.dim, self.mlp_hidden_dim]),
            get_param(f"{self.prefix}.fc2.bias", [self.dim]),
        )
        x = add(x, identity)
        return x


class VisionTransformer:
    """Vision Transformer."""

    def __init__(  # noqa: D107
        self,
        img_size: int = 224,
        patch_size: int = 16,
        in_chans: int = 3,
        num_classes: int = 1000,
        embed_dim: int = 768,
        depth: int = 12,
        num_heads: int = 12,
        mlp_ratio: float = 4.0,
        qkv_bias: bool = True,
    ):
        self.embed_dim = embed_dim
        self.num_classes = num_classes
        self.patch_embed = PatchEmbed(
            img_size=img_size,
            patch_size=patch_size,
            in_chans=in_chans,
            embed_dim=embed_dim,
            prefix="patch_embed",
        )
        self.blocks = [
            Block(
                dim=embed_dim,
                num_heads=num_heads,
                mlp_ratio=mlp_ratio,
                qkv_bias=qkv_bias,
                prefix=f"blocks.{i}",
            )
            for i in range(depth)
        ]
        self.norm = LayerNormalization((embed_dim,), epsilon=1e-6)
        self.head = Gemm(trans_b=1)
        self.num_patches = self.patch_embed.num_patches

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        x = self.patch_embed(x)

        cls_token = get_param("cls_token", [1, 1, self.embed_dim])
        # Concat cls_token along axis 1 (seq_len)
        # Note: in real models cls_token is expanded to batch size, but for AST build we just concat
        x = concat([cls_token, x], axis=1)

        pos_embed = get_param("pos_embed", [1, self.num_patches + 1, self.embed_dim])
        x = add(x, pos_embed)

        for block in self.blocks:
            x = block(x)

        x = self.norm(
            x, get_param("norm.weight", [self.embed_dim]), get_param("norm.bias", [self.embed_dim])
        )

        # Take the cls_token output (index 0)
        from onnx9000.core.ops import constant, record_op

        starts = constant([0], dtype=7)
        ends = constant([1], dtype=7)
        axes = constant([1], dtype=7)
        x = slice_op(x, starts, ends, axes)

        x = record_op("Flatten", [x], {"axis": 1})  # flatten spatial dims

        x = self.head(
            x,
            get_param("head.weight", [self.num_classes, self.embed_dim]),
            get_param("head.bias", [self.num_classes]),
        )
        return x


def vit_base_patch16_224(**kwargs: Any) -> VisionTransformer:  # noqa: D103
    return VisionTransformer(patch_size=16, embed_dim=768, depth=12, num_heads=12, **kwargs)

File: packages/js/core/src/models/vit.ts
import { Tensor } from '../ir/tensor.js';
import {
  BatchNormalization,
  ConvND,
  Gemm,
  LayerNormalization,
  MultiHeadAttention,
  Gelu,
} from '../primitives.js';

function getParam(
  name: string,
  shape: number[],
  dtype: ReturnType<typeof JSON.parse> = 'float32',
): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: ReturnType<typeof JSON.parse>): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class PatchEmbed {
  public prefix: string;
  public imgSize: number;
  public patchSize: number;
  public numPatches: number;
  public proj: ConvND;

  constructor(
    imgSize: number = 224,
    patchSize: number = 16,
    inChans: number = 3,
    embedDim: number = 768,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.imgSize = imgSize;
    this.patchSize = patchSize;
    this.numPatches = Math.floor(imgSize / patchSize) * Math.floor(imgSize / patchSize);
    this.proj = new ConvND(2, inChans, embedDim, patchSize, patchSize, 0, 1, 1, false);
  }

  call(x: Tensor): Tensor {
    x = this.proj.call(
      x,
      getParam(`${this.prefix}.proj.weight`, [
        this.proj.outChannels,
        this.proj.inChannels,
        this.patchSize,
        this.patchSize,
      ]),
      getParam(`${this.prefix}.proj.bias`, [this.proj.outChannels]),
    );
    x = recordOp('Flatten', [x], { axis: 2 });
    x = recordOp('Transpose', [x], { perm: [0, 2, 1] });
    return x;
  }
}

export class Block {
  public prefix: string;
  public dim: number;
  public norm1: LayerNormalization;
  public attn: MultiHeadAttention;
  public norm2: LayerNormalization;
  public fc1: Gemm;
  public act: Gelu;
  public fc2: Gemm;
  public mlpHiddenDim: number;

  constructor(
    dim: number,
    numHeads: number,
    mlpRatio: number = 4.0,
    qkvBias: boolean = false,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.dim = dim;
    this.norm1 = new LayerNormalization([dim], 1e-6);
    this.attn = new MultiHeadAttention(numHeads, qkvBias, true);
    this.norm2 = new LayerNormalization([dim], 1e-6);
    this.mlpHiddenDim = Math.floor(dim * mlpRatio);
    this.fc1 = new Gemm(1.0, 1.0, 0, 1);
    this.act = new Gelu();
    this.fc2 = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    let identity = x;
    x = this.norm1.call(
      x,
      getParam(`${this.prefix}.norm1.weight`, [this.dim]),
      getParam(`${this.prefix}.norm1.bias`, [this.dim]),
    );
    x = this.attn.call(x, x, x);
    x = recordOp('Add', [x, identity]);

    identity = x;
    x = this.norm2.call(
      x,
      getParam(`${this.prefix}.norm2.weight`, [this.dim]),
      getParam(`${this.prefix}.norm2.bias`, [this.dim]),
    );
    x = this.fc1.call(
      x,
      getParam(`${this.prefix}.fc1.weight`, [this.mlpHiddenDim, this.dim]),
      getParam(`${this.prefix}.fc1.bias`, [this.mlpHiddenDim]),
    );
    x = this.act.call(x);
    x = this.fc2.call(
      x,
      getParam(`${this.prefix}.fc2.weight`, [this.dim, this.mlpHiddenDim]),
      getParam(`${this.prefix}.fc2.bias`, [this.dim]),
    );
    x = recordOp('Add', [x, identity]);
    return x;
  }
}

export class VisionTransformer {
  public embedDim: number;
  public numClasses: number;
  public patchEmbed: PatchEmbed;
  public blocks: Block[];
  public norm: LayerNormalization;
  public head: Gemm;
  public numPatches: number;

  constructor(
    imgSize: number = 224,
    patchSize: number = 16,
    inChans: number = 3,
    numClasses: number = 1000,
    embedDim: number = 768,
    depth: number = 12,
    numHeads: number = 12,
    mlpRatio: number = 4.0,
    qkvBias: boolean = true,
  ) {
    this.embedDim = embedDim;
    this.numClasses = numClasses;
    this.patchEmbed = new PatchEmbed(imgSize, patchSize, inChans, embedDim, 'patch_embed');

    this.blocks = [];
    for (let i = 0; i < depth; i++) {
      this.blocks.push(new Block(embedDim, numHeads, mlpRatio, qkvBias, `blocks.${i}`));
    }

    this.norm = new LayerNormalization([embedDim], 1e-6);
    this.head = new Gemm(1.0, 1.0, 0, 1);
    this.numPatches = this.patchEmbed.numPatches;
  }

  call(x: Tensor): Tensor {
    x = this.patchEmbed.call(x);

    const clsToken = getParam('cls_token', [1, 1, this.embedDim]);
    x = recordOp('Concat', [clsToken, x], { axis: 1 });

    const posEmbed = getParam('pos_embed', [1, this.numPatches + 1, this.embedDim]);
    x = recordOp('Add', [x, posEmbed]);

    for (const block of this.blocks) {
      x = block.call(x);
    }

    x = this.norm.call(
      x,
      getParam('norm.weight', [this.embedDim]),
      getParam('norm.bias', [this.embedDim]),
    );

    const starts = recordOp('Constant', [], { value: [0], dtype: 7 });
    const ends = recordOp('Constant', [], { value: [1], dtype: 7 });
    const axes = recordOp('Constant', [], { value: [1], dtype: 7 });
    x = recordOp('Slice', [x, starts, ends, axes]);

    x = recordOp('Flatten', [x], { axis: 1 });

    x = this.head.call(
      x,
      getParam('head.weight', [this.numClasses, this.embedDim]),
      getParam('head.bias', [this.numClasses]),
    );
    return x;
  }
}

export function vitBasePatch16_224(numClasses: number = 1000): VisionTransformer {
  return new VisionTransformer(224, 16, 3, numClasses, 768, 12, 12, 4.0, true);
}
